standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -1236,31 +1237,60 @@ public void replTableWriteIdState(ReplTblWriteIdStateRequest rqst) throws MetaEx
     return abortedWriteIds;
   }
 
+  private ValidTxnList getValidTxnList(Connection dbConn, String fullTableName, Long writeId) throws MetaException,
+          SQLException {
+    PreparedStatement pst = null;
+    ResultSet rs = null;
+    try {
+      String[] names = TxnUtils.getDbTableName(fullTableName);
+      String s = "select t2w_txnid from TXN_TO_WRITE_ID where  t2w_database = ? and t2w_table = ? and t2w_writeid = ?";
+      pst = dbConn.prepareStatement(sqlGenerator.addEscapeCharacters(s));
+      pst.setString(1, names[0]);
+      pst.setString(2, names[1]);
+      pst.setLong(3, writeId);
+      LOG.debug("Going to execute query <" + s.replaceAll("\\?", "{}") + ">", quoteString(names[0]),
+              quoteString(names[1]), writeId);
+      rs = pst.executeQuery();
+      if (rs.next()) {
+        return TxnCommonUtils.createValidReadTxnList(getOpenTxns(), rs.getLong(1));
+      }
+      throw new MetaException("invalid write id " + writeId + " for table " + fullTableName);
+    } finally {
+      close(rs);
+      if (pst != null) {
+        pst.close();
```

Comment:
Shall call closeStmt(pst) and avoid null check here.

Reply:
clostStmt is not there for prepared stmts

Reply:
done